[JENKINS-48227] Use "Files.createTempDirectory" to create temp directory

larrysingleton opened the pull request:
Jira contains link to sonar squid number as well as link to personal sonar cloud analysis report relating to this issue

larrysingleton committed:
Merge pull request #1 from jenkinsci/master

Sync Up 10-03-2017

larrysingleton committed:
Use "Files.createTempDirectory" to create temp directory instead
See SonarQube critical vulnerability squid:S2976 (tag: owasp-a9)
https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2976

oleg-nenashev commented on core/src/main/java/hudson/FilePath.java:
```diff
@@ -1400,17 +1402,15 @@ public String invoke(File dir, VirtualChannel channel) throws IOException {
-                    f.delete();
-                    f.mkdir();
-                    return f.getName();
+                    Path tempPath = Files.createTempDirectory(dir.toPath(), prefix, null);
```
`dir.toPath()` may throw `InvalidPathException`. Since it is an API method, it has to be handled and wrapped to `IOException`

larrysingleton replied:
Agreed - change added

oleg-nenashev reviewed (changes requested):
See the comment. I would suggest creating an Utility `static Path toPath(File file) throws IOException` method, which wraps `InvalidPathException` to `IOException`so that it will be checked.

larrysingleton committed:
[JENKINS-48227]
Creating a utility "static Path toPath(File file) throws IOException" method,
which wraps InvalidPathException to IOException so that it will be checked.
- also fixed public static final reference
- fixed broken test cases
- added new test cases for toPath() and createTempDir()

larrysingleton commented:
Suggested changes have been added.

oleg-nenashev commented on core/src/main/java/hudson/FilePath.java:
```diff
@@ -2371,7 +2402,7 @@ public String validateAntFileMask(final String fileMasks, final int bound) throw
      * @since 1.592
      */
-    public static int VALIDATE_ANT_FILE_MASK_BOUND = SystemProperties.getInteger(FilePath.class.getName() + ".VALIDATE_ANT_FILE_MASK_BOUND", 10000);
+    public static final int VALIDATE_ANT_FILE_MASK_BOUND = SystemProperties.getInteger(FilePath.class.getName() + ".VALIDATE_ANT_FILE_MASK_BOUND", 10000);
```
This change is unrelated && it causes merge conflict.
Note that some fields are bob-final for a reason, so that they cannot be modified in System scripts

larrysingleton replied:
reverted

larrysingleton committed:
Revert back to public static int

oleg-nenashev committed:
Merge branch 'master' into master

oleg-nenashev commented:
Resolved the merge conflict

larrysingleton commented:
@oleg-nenashev - is there anything else you need from me here?
The test case failure does not appear to be related.
Let me know otherwise.

dwnusbaum commented:
Along with resolving the conflicts it might make sense to get rid of the newly added `FilePath#toPath` and use `Util#fileToPath` now that #3135 is merged.

larrysingleton committed:
Merge branch 'master' of https://github.com/jenkinsci/jenkins

larrysingleton committed:
adjustments due to merges

dwnusbaum reviewed (approved):
I have a minor comment but I think the PR is good as-is.

larrysingleton commented on core/src/main/java/hudson/FilePath.java:
```diff
@@ -1400,17 +1403,25 @@ public String invoke(File dir, VirtualChannel channel) throws IOException {
+                    if (tempPath.toFile() == null) {
+                        throw new IOException("Failed to obtain file from path " + dir + " on " + remote);
+                    }
+                    return tempPath.toFile().getName();
```
tempPath.getFileName() is not null safe either - a check would still be necessary
https://docs.oracle.com/javase/7/docs/api/java/nio/file/Path.html#getFileName()

Happy to make the change either way - please advise.

dwnusbaum replied:
No, I think the current code is good. My thought was that we had enough information here to know that the path is non-empty and so the call would not return null, but maybe that's not the case.

dwnusbaum commented on core/src/main/java/hudson/FilePath.java:
```diff
@@ -1400,17 +1403,25 @@ public String invoke(File dir, VirtualChannel channel) throws IOException {
-                    f.delete();
-                    f.mkdir();
-                    return f.getName();
+                    Path tempPath = Files.createTempDirectory(Util.fileToPath(dir), name,  new FileAttribute<?>[] {});
```
Found a potentially relevant issue while investigating [JENKINS-48407](https://issues.jenkins-ci.org/browse/JENKINS-48407). The old io code would create directories with permissions `0777 & (~umask)`, whereas the nio code will create directories with permissions `0700 & (~umask)`. It's easy to retain the old behavior by passing `PosixFileAttributes.asFileAttributes(EnumSet.allOf(PosixFilePermission.class))` as the file attribute parameter to `Files#createTemporaryDirectory` after checking `Functions#isWindows`. CC @oleg-nenashev do you think it is important to maintain the old permissions here?

oleg-nenashev replied:
Better safe than sorry. I would vote for retaining the permissions in this case for sure

dwnusbaum commented:
Also I just noticed that `hudson.model.Util#createTempDir` appears to have the same issue, do you want to update it in this PR as well?

oleg-nenashev reviewed (changes requested):
Needs a fix of the default permissions change (see the comment from @dwnusbaum )

larrysingleton committed:
Add posix check to determine if Posix FileAttributes should be included
in call to Files.createTempDirectory()

larrysingleton committed:
Merge branch 'master' of https://github.com/jenkinsci/jenkins

larrysingleton commented:
Add posix check as requested

larrysingleton committed:
Remove reference to private element

larrysingleton commented:
@oleg-nenashev / @dwnusbaum 
Requested changes have been added, build passes checks with no conflicts

dwnusbaum commented on core/src/main/java/hudson/FilePath.java:
```diff
@@ -77,16 +77,11 @@
-import java.util.List;
-import java.util.Map;
-import java.util.StringTokenizer;
+import java.nio.file.*;
```
nit: I prefer to avoid `.*` imports to make it clear exactly what classes are being used.

dwnusbaum commented on core/src/main/java/hudson/FilePath.java:
```diff
@@ -77,16 +77,11 @@
+import java.nio.file.attribute.FileAttribute;
+import java.nio.file.attribute.PosixFilePermission;
+import java.nio.file.attribute.PosixFilePermissions;
+import java.util.*;
```
Same

dwnusbaum reviewed (approved):
This looks good to me, although I would prefer to make the `.*` imports explicit again.

larrysingleton committed:
Updated to use explicit imports

larrysingleton commented:
@dwnusbaum - Updated to use explicit imports

oleg-nenashev committed:
Merge branch 'master' into master

oleg-nenashev commented:
Fixed the merge conflict

oleg-nenashev commented on core/src/test/java/hudson/FilePathTest.java:
```diff
@@ -641,7 +644,7 @@ public void testValidateCaseSensitivity() throws Exception {
               .thenReturn(someZippedContent());
 
-            assertTrue(d.installIfNecessaryFrom(url, null, null));
+            assertTrue(d.installIfNecessaryFrom(url, null, ""));
```
Does not seem to be related

oleg-nenashev replied:
The annotation has been introduced by @jglick in https://github.com/jenkinsci/jenkins/blame/09bcc5d6538b3cfffbf71228ebd1679e3e20d8b2/core/src/main/java/hudson/FilePath.java#L766 3 years ago, Does not matter now IMHO, I am fine with fixing it in this PR

larrysingleton replied:
True, it is not directly related - however the test case fails without the change.

oleg-nenashev replied:
@daniel-beck @jglick @dwnusbaum do we care about that? Maybe we need to do some extra NPE checks

oleg-nenashev reviewed (approved):
LGTM
